### ISSUE TITLE
Plugin to manage and integrate data loaded from an API

### ISSUE THREAD

**Issue by stefandesu** [gbv/jskos-vue#7]
In Cocoda, we have the so called ["CDK mixin"](https://github.com/gbv/cocoda/blob/a23b7820abe2d721f3d975ede49d357b98737eb2/src/mixins/cdk.js). This is a wrapper around cocoda-sdk which deals with pretty much all loading of data inside of Cocoda. There are many reasons for having something like this:

- We can reduce the number of requests by batching them and caching results so that already loaded data is not immediately requested again.
- We can connect the data we loaded so that it's easier to pass data around. For example, the `inScheme` property of concepts should contain a reference to that concept scheme, and if all concepts of that scheme use the reference, we can always get the most current object. Same for narrower and ancestor concepts, mapping members, and so on.
- We can make sure that the data that comes from the API conforms to certain standards, e.g. items should have a `type` property so that we can easily differentiate between concepts and schemes.
- There is one central place in the application that deals with loading data from the API.
- It can also connect events from the API (for example errors, successful POST request, ...) to the interface, or at least provide a defined way to catch those events so that the application can decide whether or not to show something to the user.

As you can see, there are lots of reasons for having such a plugin, but it's also fairly complex. There are also some things that the "CDK mixin" does not do well. For example, every single thing that was once loaded from the API gets kept in memory forever. There is (currently) no way of detecting which items are actively in use and which items haven't been accessed in some time, so we can't easily clear data from memory. I haven't yet found a solution for this. The consequence is that if you use Cocoda for a long period of time, it will use more and more memory. Fortunately, the data is fairly small, so this hasn't been an issue yet, but eventually this should be solved somehow.

I'd be very happy for any thoughts on this issue. Should this be part of jskos-vue, or maybe a separate package? Have I forgotten anything that we also need in terms of functionality? Any other thoughts or ideas?

**Comment by nichtich** [comment]
By now the only component that queries a backend API is ItemSuggest. I'd prefer to get all API interaction out of jskos-vue and use plain promises instead, so data loading is decoupled from jskos-vue (see gbv/jskos-vue#14).

**Comment by stefandesu** [comment]
I mean I did suggest to put it in a separate package, so all my points still stand, and I agree that it make sense to keep API interaction out of jskos-vue and focus it on interface components.

I'm not sure what you mean by "use plain promises instead" though.

**Comment by nichtich** [comment]
> I'm not sure what you mean by "use plain promises instead" though.

If a component of jskos-vue should dynamically load content, then this should be done via a property that is given an async function such as `search` in `ItemSuggest`. How this functions gets data is out of the scope of jskos-vue.

**Comment by stefandesu** [comment]
> If a component of jskos-vue should dynamically load content, then this should be done via a property that is given an async function such as `search` in `ItemSuggest`. How this functions gets data is out of the scope of jskos-vue.

Yeah that's totally fine with me and makes sense. But it doesn't have anything to do with this issue. 😅
